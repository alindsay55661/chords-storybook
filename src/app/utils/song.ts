import type { Scale } from '@tonaljs/scale'
import type { BaseSongData, Bar, Track, Note } from './parse'

export type DetectUnit = 'bar' | 'beat'

export type NoteStats = {
  frequency: Record<string, number>
  duration: Record<string, number>
  presence?: Record<string, number>
  map: Record<string, Note>
  byBar: Bar[]
}

export type Song = Pick<BaseSongData, 'timings' | 'timeSignatures'> & {
  notes: NoteStats
  scales: Scale[]
  tracks: Track[]
}

export function makeSong(sd: BaseSongData): Song {
  let song: Song = {
    timings: sd.timings,
    timeSignatures: sd.timeSignatures,
    notes: {
      frequency: {},
      duration: {},
      presence: {},
      map: {},
      byBar: [],
    },
    scales: [],
    tracks: [],
  }

  // Execution order matters
  // 1 - every note must be analyzed before scale or chord detection can occur
  // 2 - note "presence" improve scale recognition and should be done first
  // 3 - scale recognition helps with proper chord spelling
  // 4 - chord recognition is the last thing to take place

  // 1 - Note analysis (Hydrate barsAndBeats)
  const notes = Object.values(sd.notesMap)

  sd.barsAndBeats.forEach(bar => {
    // Find notes in this bar
    const barEndTicks = bar.startTicks + bar.durationTicks
    const notesInBar = notes.filter(note => {
      if (isNoteInRange(note, bar.startTicks, barEndTicks)) return note
    })
    bar.notes = notesInBar

    // Find notes in this bar's beats
    bar.beats.forEach(beat => {
      const beatEndTicks = beat.startTicks + beat.durationTicks
      const notesInBeat = notesInBar.filter(note => {
        if (isNoteInRange(note, beat.startTicks, beatEndTicks)) return note
      })
      beat.notes = notesInBeat
    })
  })

  song.notes.byBar = sd.barsAndBeats
  song.tracks = sd.tracks
  // song.notes.frequency = updateFrequencyStats(note, song)
  // song.notes.duration = updateDurationStats(note, song)

  // 2 - Note "presence" calculation
  song.notes.presence = updatePresenceStats(song) as Record<string, number>

  // 3-4 are done with other funciton calls
  return song
}

export function isNoteInRange(
  note: Note,
  startTicks: number,
  endTicks: number,
) {
  const noteEndTicks = note.startTicks + note.durationTicks

  // Note starts and ends in the range (CONTAINED - most common)
  if (note.startTicks >= startTicks && note.startTicks <= endTicks) {
    return true
  }

  // Note starts in the range and ends after the range (HEAD)
  if (note.startTicks >= startTicks && note.startTicks <= endTicks) {
    return true
  }

  // Note starts before the range and ends in the range (TAIL)
  if (noteEndTicks >= startTicks && noteEndTicks <= endTicks) {
    return true
  }

  // Note starts before the range and ends after range (THROUGH)
  if (note.startTicks < startTicks && noteEndTicks > endTicks) {
    return true
  }

  return false
}

function updateFrequencyStats(note: Note, song: Song) {
  const name = note.noteName[0]
  const frequency = song.notes.frequency ?? {}
  frequency[name] = frequency[name] ?? 0
  frequency[name] = frequency[name] + 1
  return frequency
}

function updateDurationStats(note: Note, song: Song) {
  const name = note.noteName[0]
  const duration = song.notes.duration ?? {}
  duration[name] = duration[name] ?? 0
  duration[name] = duration[name] + note.durationTicks
  return duration
}

export function updatePresenceStats(
  song: Song,
  noteDurationWeight: number = 1,
  noteFrequencyWeight: number = 0.3,
) {
  const presence = song.notes.presence ?? {}
  const duration = normalizeToOne(song.notes.duration)
  const frequency = normalizeToOne(song.notes.frequency)

  Object.keys(frequency).forEach(name => {
    presence[name] =
      duration[name] * noteDurationWeight +
      frequency[name] * noteFrequencyWeight
  })

  return normalizeToOne(presence)
}

function normalizeToOne(noteMap: Record<string, number>) {
  const values = Object.values(noteMap)
  const min = 0 //Math.min(...values)
  const max = Math.max(...values)
  const factor = 1 / (max - min)

  return Object.entries(noteMap).reduce((result: any, tuple) => {
    result[tuple[0]] = (tuple[1] - min) * factor
    return result
  }, {})
}
