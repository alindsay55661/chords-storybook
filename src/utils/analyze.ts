import type { Scale } from '@tonaljs/scale'
import type { MidiPart, Note } from './parse'

export type DetectUnit = 'bar' | 'beat'

export type Beat = {
  index: number
  startTicks: number
  notes: Note[]
}

export type NoteStats = {
  frequency: Record<string, number>
  duration: Record<string, number>
  presence?: Record<string, number>
  map: Record<string, Note>
  byBeat: Beat[]
}

export type Stats = Pick<MidiPart, 'timings' | 'timeSignatures'> & {
  notes: NoteStats
  scales: Scale[]
}

export function analyze(part: MidiPart): Stats {
  let stats: Stats = {
    timings: part.timings,
    timeSignatures: part.timeSignatures,
    notes: {
      frequency: {},
      duration: {},
      presence: {},
      map: {},
      byBeat: new Array(part.timings.totalBeats),
    },
    scales: [],
  }

  // Execution order matters
  // 1 - every note must be analyzed before scale or chord detection can occur
  // 2 - note "presence" improve scale recognition and should be done first
  // 3 - scale recognition helps with proper chord spelling
  // 4 - chord recognition is the last thing to take place

  // 1 - Note analysis
  part.tracks.forEach(track => {
    track.notes.forEach(note => {
      stats = analyzeNote(note, stats)
    })
  })

  // 2 - Note "presence" calculation
  stats.notes.presence = updatePresenceStats(stats) as Record<string, number>

  // 3-4 are done with other funciton calls
  return stats
}

// Adds this note to the current stats
function analyzeNote(note: Note, stats: Stats) {
  // Add note stats
  stats.notes.frequency = updateFrequencyStats(note, stats)
  stats.notes.duration = updateDurationStats(note, stats)
  stats.notes.map[note.uuid] = note

  const noteOnTicks = note.startTicks
  const noteOffTicks = note.startTicks + note.durationTicks
  const noteStartRange = Math.floor(noteOnTicks / stats.timings.ticksPerBeat)
  const noteEndRange = Math.floor(noteOffTicks / stats.timings.ticksPerBeat)

  // loop is limited to a note's own buckets (typically 1-2)
  for (let i = noteStartRange; i <= noteEndRange; i++) {
    const beat = stats.notes.byBeat[i] ?? {
      index: i,
      startTicks: i * stats.timings.ticksPerBeat,
      notes: [],
    }

    beat.notes.push(note)
    stats.notes.byBeat[i] = beat
  }

  return stats
}

function updateFrequencyStats(note: Note, stats: Stats) {
  const name = note.noteName[0]
  const frequency = stats.notes.frequency ?? {}
  frequency[name] = frequency[name] ?? 0
  frequency[name] = frequency[name] + 1
  return frequency
}

function updateDurationStats(note: Note, stats: Stats) {
  const name = note.noteName[0]
  const duration = stats.notes.duration ?? {}
  duration[name] = duration[name] ?? 0
  duration[name] = duration[name] + note.durationTicks
  return duration
}

export function updatePresenceStats(
  stats: Stats,
  noteDurationWeight: number = 1,
  noteFrequencyWeight: number = 0.3,
) {
  const presence = stats.notes.presence ?? {}
  const duration = normalizeToOne(stats.notes.duration)
  const frequency = normalizeToOne(stats.notes.frequency)

  Object.keys(frequency).forEach(name => {
    presence[name] =
      duration[name] * noteDurationWeight +
      frequency[name] * noteFrequencyWeight
  })

  return normalizeToOne(presence)
}

function normalizeToOne(noteMap: Record<string, number>) {
  const values = Object.values(noteMap)
  const min = 0 //Math.min(...values)
  const max = Math.max(...values)
  const factor = 1 / (max - min)

  return Object.entries(noteMap).reduce((result: any, tuple) => {
    result[tuple[0]] = (tuple[1] - min) * factor
    return result
  }, {})
}

export function noteInBeat(note: Note, beat: number) {
  return true
}
