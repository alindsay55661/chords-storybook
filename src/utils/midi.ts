import Tonal from 'tonal'
import { v4 as uuid } from 'uuid'
import { buildMidiPart } from './adapters/midi-file-adapter'

export type MidiPart = {
  timings: Timings
  timeSignature: TimeSignature
  tracks: Track[]
}

export type Timings = {
  durationTicks: number
  ticksPerBeat: number
  microsecondsPerBeat: number
}

export type TimeSignature = {
  numerator: number
  denominator: number
}

export type Track = {
  notes: Note[]
  durationTicks: number
  name?: string
  midiPatch?: number
}

export type Note = {
  uuid: string
  startTicks: number
  durationTicks: number
  noteNumber: number
  noteNameWithOctave: string[]
  noteName: string[]
}

export function parseMidi(data: ArrayLike<number>): MidiPart {
  return buildMidiPart(data)
}

export function buildNote(
  note: Pick<Note, 'startTicks' | 'noteNumber'>,
  partStartTicks: number,
): Note | false {
  const durationTicks = partStartTicks - note.startTicks

  // Ignore notes with no duration (this happens with some midi programs)
  if (!durationTicks) return false

  return {
    uuid: uuid(),
    startTicks: partStartTicks,
    durationTicks,
    noteNumber: note.noteNumber,
    noteNameWithOctave: [
      Tonal.Midi.midiToNoteName(note.noteNumber),
      Tonal.Midi.midiToNoteName(note.noteNumber, { sharps: true }),
    ],
    noteName: [
      Tonal.Midi.midiToNoteName(note.noteNumber, { pitchClass: true }),
      Tonal.Midi.midiToNoteName(note.noteNumber, {
        pitchClass: true,
        sharps: true,
      }),
    ],
  }
}
type GMPatchNames = {
  general: Record<number, string>
  percussion: Record<number, string>
}

export const GMPatchNames: GMPatchNames = {
  general: {
    0: 'Acoustic Grand',
    1: 'Bright Acoustic',
    2: 'Electric Grand',
    3: 'Honky-Tonk',
    4: 'Electric Piano 1',
    5: 'Electric Piano 2',
    6: 'Harpsichord',
    7: 'Clav',
    8: 'Celesta',
    9: 'Glockenspiel',
    10: 'Music Box',
    11: 'Vibraphone',
    12: 'Marimba',
    13: 'Xylophone',
    14: 'Tubular Bells',
    15: 'Dulcimer',
    16: 'Drawbar Organ',
    17: 'Percussive Organ',
    18: 'Rock Organ',
    19: 'Church Organ',
    20: 'Reed Organ',
    21: 'Accordion',
    22: 'Harmonica',
    23: 'Tango Accordion',
    24: 'Acoustic Guitar(nylon)',
    25: 'Acoustic Guitar(steel)',
    26: 'Electric Guitar(jazz)',
    27: 'Electric Guitar(clean)',
    28: 'Electric Guitar(muted)',
    29: 'Overdriven Guitar',
    30: 'Distortion Guitar',
    31: 'Guitar Harmonics',
    32: 'Acoustic Bass',
    33: 'Electric Bass(finger)',
    34: 'Electric Bass(pick)',
    35: 'Fretless Bass',
    36: 'Slap Bass 1',
    37: 'Slap Bass 2',
    38: 'Synth Bass 1',
    39: 'Synth Bass 2',
    40: 'Violin',
    41: 'Viola',
    42: 'Cello',
    43: 'Contrabass',
    44: 'Tremolo Strings',
    45: 'Pizzicato Strings',
    46: 'Orchestral Harp',
    47: 'Timpani',
    48: 'String Ensemble 1',
    49: 'String Ensemble 2',
    50: 'SynthStrings 1',
    51: 'SynthStrings 2',
    52: 'Choir Aahs',
    53: 'Voice Oohs',
    54: 'Synth Voice',
    55: 'Orchestra Hit',
    56: 'Trumpet',
    57: 'Trombone',
    58: 'Tuba',
    59: 'Muted Trumpet',
    60: 'French Horn',
    61: 'Brass Section',
    62: 'SynthBrass 1',
    63: 'SynthBrass 2',
    64: 'Soprano Sax',
    65: 'Alto Sax',
    66: 'Tenor Sax',
    67: 'Baritone Sax',
    68: 'Oboe',
    69: 'English Horn',
    70: 'Bassoon',
    71: 'Clarinet',
    72: 'Piccolo',
    73: 'Flute',
    74: 'Recorder',
    75: 'Pan Flute',
    76: 'Blown Bottle',
    77: 'Shakuhachi',
    78: 'Whistle',
    79: 'Ocarina',
    80: 'Lead 1 (square)',
    81: 'Lead 2 (sawtooth)',
    82: 'Lead 3 (calliope)',
    83: 'Lead 4 (chiff)',
    84: 'Lead 5 (charang)',
    85: 'Lead 6 (voice)',
    86: 'Lead 7 (fifths)',
    87: 'Lead 8 (bass+lead)',
    88: 'Pad 1 (new age)',
    89: 'Pad 2 (warm)',
    90: 'Pad 3 (polysynth)',
    91: 'Pad 4 (choir)',
    92: 'Pad 5 (bowed)',
    93: 'Pad 6 (metallic)',
    94: 'Pad 7 (halo)',
    95: 'Pad 8 (sweep)',
    96: 'FX 1 (rain)',
    97: 'FX 2 (soundtrack)',
    98: 'FX 3 (crystal)',
    99: 'FX 4 (atmosphere)',
    100: 'FX 5 (brightness)',
    101: 'FX 6 (goblins)',
    102: 'FX 7 (echoes)',
    103: 'FX 8 (sci-fi)',
    104: 'Sitar',
    105: 'Banjo',
    106: 'Shamisen',
    107: 'Koto',
    108: 'Kalimba',
    109: 'Bagpipe',
    110: 'Fiddle',
    111: 'Shanai',
    112: 'Tinkle Bell',
    113: 'Agogo',
    114: 'Steel Drums',
    115: 'Woodblock',
    116: 'Taiko Drum',
    117: 'Melodic Tom',
    118: 'Synth Drum',
    119: 'Reverse Cymbal',
    120: 'Guitar Fret Noise',
    121: 'Breath Noise',
    122: 'Seashore',
    123: 'Bird Tweet',
    124: 'Telephone Ring',
    125: 'Helicopter',
    126: 'Applause',
    127: 'Gunshot',
  },
  percussion: {
    24: 'Snare Roll',
    25: 'Finger Snap',
    26: 'High Q',
    27: 'Slap',
    28: 'Scratch Pull',
    29: 'Scratch Push',
    30: 'Sticks',
    31: 'Square Click',
    32: 'Metronome Bell',
    33: 'Metronome Click',
    34: 'Acoustic Bass Drum',
    35: 'Electric Bass Drum',
    36: 'Side Stick',
    37: 'Acoustic Snare',
    38: 'Hand Clap',
    39: 'Electric Snare',
    40: 'Low Floor Tom',
    41: 'Closed Hi-hat',
    42: 'High Floor Tom',
    43: 'Pedal Hi-hat',
    44: 'Low Tom',
    45: 'Open Hi-hat',
    46: 'Low-Mid Tom',
    47: 'High-Mid Tom',
    48: 'Crash Cymbal 1',
    49: 'High Tom',
    50: 'Ride Cymbal 1',
    51: 'Chinese Cymbal',
    52: 'Ride Bell',
    53: 'Tambourine',
    54: 'Splash Cymbal',
    55: 'Cowbell',
    56: 'Crash Cymbal 2',
    57: 'Vibraslap',
    58: 'Ride Cymbal 2',
    59: 'High Bongo',
    60: 'Low Bongo',
    61: 'Mute High Conga',
    62: 'Open High Conga',
    63: 'Low Conga',
    64: 'High Timbale',
    65: 'Low Timbale',
    66: 'High Agogô',
    67: 'Low Agogô',
    68: 'Cabasa',
    69: 'Maracas',
    70: 'Short Whistle',
    71: 'Long Whistle',
    72: 'Short Guiro',
    73: 'Long Guiro',
    74: 'Claves',
    75: 'High Woodblock',
    76: 'Low Woodblock',
    77: 'Mute Cuica',
    78: 'Open Cuica',
    79: 'Mute Triangle',
    80: 'Open Triangle',
    81: 'Shaker',
    82: 'Jingle Bell',
    83: 'Belltree',
    84: 'Castanets',
    85: 'Mute Surdo',
    86: 'Open Surdo',
  },
}
