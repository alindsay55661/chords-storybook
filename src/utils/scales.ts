import Tonal from 'tonal'

export function detectScales(presence: Record<string, number>) {
  // order by prominence
  const notes = Object.entries(presence)
    .map(entry => {
      return { name: entry[0], presence: entry[1] }
    })
    .sort((a, b) => {
      if (a.presence > b.presence) return -1
      if (a.presence < b.presence) return 1
      return 0
    })

  const scaleNotes = Tonal.Scale.scaleNotes(notes.map(note => note.name))

  // Get distance from C
  const interval = Tonal.Interval.distance(scaleNotes[0], 'C')
  const transposed = scaleNotes.map(Tonal.Note.transposeBy(interval))
  const pcset = Tonal.Pcset.get(transposed)

  // Filter scales that include these intervals
  // https://github.com/tonaljs/tonal/blob/main/packages/chord/index.ts#L197
  const areNotesIncluded = Tonal.Pcset.isSupersetOf(pcset.chroma)
  const scales = Tonal.ScaleType.all().filter(scale =>
    areNotesIncluded(scale.chroma),
  )

  const keyScales = scales.map(scale => {
    return Tonal.Scale.get(`${scaleNotes[0]} ${scale.name}`)
  })

  return keyScales
}
