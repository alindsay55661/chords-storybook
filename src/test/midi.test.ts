import { expect, test } from 'vitest'
import { readFileSync } from 'node:fs'
import { parseMidi, Note } from '../utils/midi'
import {
  detectChords,
  updateDistribution,
  Bucket,
} from '../utils/music-analysis'

test('parseMidi() should parse to a known format', () => {
  const path = `${__dirname}/beat.mid`
  const data = readFileSync(path)
  const parsed = parseMidi(data)

  // Do some light schema validation
  expect(parsed.timeSignature.denominator).toBeDefined()
  expect(parsed.timeSignature.numerator).toBeDefined()
  expect(parsed.timings.durationTicks).toBeDefined()
  expect(parsed.timings.microsecondsPerBeat).toBeDefined()
  expect(parsed.timings.ticksPerBeat).toBeDefined()
  expect(parsed.tracks[0].notes[0].durationTicks).toBeDefined()
  expect(parsed.tracks[0].notes[0].noteNameWithOctave).toBeDefined()
  expect(parsed.tracks[0].notes[0].noteName).toBeDefined()
  expect(parsed.tracks[0].notes[0].noteNumber).toBeDefined()
  expect(parsed.tracks[0].notes[0].startTicks).toBeDefined()
  expect(parsed.tracks[0].durationTicks).toBe(479)
  expect(parsed.tracks[0].notes[0].noteNameWithOctave).toEqual(['A3', 'A3'])
  expect(parsed.tracks[0].notes[2].noteNameWithOctave).toEqual(['Bb3', 'A#3'])
  expect(parsed.tracks[0].notes[0].noteName).toEqual(['A', 'A'])
  expect(parsed.tracks[0].notes[2].noteName).toEqual(['Bb', 'A#'])

  // Keep a snapshot around in case adaptors or 3rd party libs are changed
  expect(parsed).toMatchSnapshot()
})

// Consider moving detection (or bucket building) to the parsing stage
test('chord detection', () => {
  const path = `${__dirname}/sample.mid`
  const data = readFileSync(path)
  const parsed = parseMidi(data)

  const chords = detectChords(parsed, { unit: 'beat' })
  console.log(chords)
})

test('updateDistribution()', () => {
  const bucket: Bucket = {
    notes: new Set(['A', 'B', 'C', 'D']),
    chords: [],
    unit: 'bar',
    unitNumber: 0,
    distribution: {
      ticks: {},
    },
    startTicks: 100,
    durationTicks: 100,
  }

  // Starts outside bucket, ends in bucket
  const noteA: Note = {
    startTicks: 80,
    durationTicks: 30,
    noteNumber: 57,
    noteName: ['A'],
    noteNameWithOctave: ['A3'],
  }

  // Starts in bucket, ends in bucket
  const noteB: Note = {
    startTicks: 120,
    durationTicks: 30,
    noteNumber: 59,
    noteName: ['B'],
    noteNameWithOctave: ['B3'],
  }

  // Starts in bucket, ends outside bucket
  const noteC: Note = {
    startTicks: 180,
    durationTicks: 30,
    noteNumber: 60,
    noteName: ['C'],
    noteNameWithOctave: ['C4'],
  }

  // Starts before bucket, ends after bucket
  const noteD: Note = {
    startTicks: 80,
    durationTicks: 130,
    noteNumber: 62,
    noteName: ['D'],
    noteNameWithOctave: ['D4'],
  }

  // Starts after bucket, ends after bucket
  const noteE: Note = {
    startTicks: 280,
    durationTicks: 10,
    noteNumber: 64,
    noteName: ['E'],
    noteNameWithOctave: ['E4'],
  }

  // Starts before bucket, ends before bucket
  const noteF: Note = {
    startTicks: 50,
    durationTicks: 10,
    noteNumber: 65,
    noteName: ['F'],
    noteNameWithOctave: ['F4'],
  }

  expect(updateDistribution(bucket, noteA).distribution).toEqual({
    ticks: {
      A: 10,
    },
  })
  expect(updateDistribution(bucket, noteB).distribution).toEqual({
    ticks: {
      A: 10,
      B: 30,
    },
  })
  expect(updateDistribution(bucket, noteC).distribution).toEqual({
    ticks: {
      A: 10,
      B: 30,
      C: 20,
    },
  })
  expect(updateDistribution(bucket, noteD).distribution).toEqual({
    ticks: {
      A: 10,
      B: 30,
      C: 20,
      D: 100,
    },
  })
  expect(updateDistribution(bucket, noteE).distribution).toEqual({
    ticks: {
      A: 10,
      B: 30,
      C: 20,
      D: 100,
    },
  })
  expect(updateDistribution(bucket, noteF).distribution).toEqual({
    ticks: {
      A: 10,
      B: 30,
      C: 20,
      D: 100,
    },
  })

  // Add A a 2nd time should increase ticks & percentage
  expect(updateDistribution(bucket, noteA).distribution).toEqual({
    ticks: {
      A: 20,
      B: 30,
      C: 20,
      D: 100,
    },
  })

  expect(updateDistribution(bucket).distribution).toEqual({
    ticks: {
      A: 20,
      B: 30,
      C: 20,
      D: 100,
    },
    time: {
      A: 0.2,
      B: 0.3,
      C: 0.2,
      D: 1,
    },
  })
})
