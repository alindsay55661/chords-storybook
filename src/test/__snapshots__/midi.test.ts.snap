// Vitest Snapshot v1

exports[`analyze() 1`] = `
{
  "chords": {
    "chordRangeCount": 4,
    "chordRanges": [
      {
        "chords": [],
        "distribution": {
          "ticks": {
            "A": 369,
            "B": 15,
            "C": 123,
            "D": 49,
            "E": 175,
            "G": 144,
          },
        },
        "durationTicks": 480,
        "notes": Set {
          "G",
          "A",
          "E",
          "C",
          "D",
          "B",
        },
        "startTicks": 0,
        "uniqueNotes": [],
        "unit": "beat",
        "unitNumber": 0,
      },
      {
        "chords": [],
        "distribution": {
          "ticks": {
            "A": 302,
            "B": 82,
            "C": 232,
            "D": 55,
            "E": 130,
            "G": 41,
          },
        },
        "durationTicks": 480,
        "notes": Set {
          "A",
          "E",
          "C",
          "G",
          "B",
          "D",
        },
        "startTicks": 480,
        "uniqueNotes": [],
        "unit": "beat",
        "unitNumber": 1,
      },
      {
        "chords": [],
        "distribution": {
          "ticks": {
            "A": 159,
            "C": 85,
            "D": 19,
            "E": 217,
            "G": 137,
          },
        },
        "durationTicks": 480,
        "notes": Set {
          "G",
          "E",
          "A",
          "C",
          "D",
        },
        "startTicks": 960,
        "uniqueNotes": [],
        "unit": "beat",
        "unitNumber": 2,
      },
      {
        "chords": [],
        "distribution": {
          "ticks": {
            "A": 197,
            "B": 167,
            "C": 310,
            "D": 200,
            "E": 47,
            "G": 104,
          },
        },
        "durationTicks": 480,
        "notes": Set {
          "E",
          "A",
          "C",
          "D",
          "G",
          "B",
        },
        "startTicks": 1440,
        "uniqueNotes": [],
        "unit": "beat",
        "unitNumber": 3,
      },
      {
        "chords": [],
        "distribution": {
          "ticks": {
            "E": 3,
          },
        },
        "durationTicks": 480,
        "notes": Set {
          "E",
        },
        "startTicks": 1920,
        "uniqueNotes": [],
        "unit": "beat",
        "unitNumber": 4,
      },
    ],
    "ticksPerChordRange": 480,
    "unit": "beat",
  },
  "notes": {
    "duration": {
      "A": 1027,
      "B": 264,
      "C": 750,
      "D": 323,
      "E": 572,
      "G": 426,
    },
    "frequency": {
      "A": 10,
      "B": 2,
      "C": 2,
      "D": 5,
      "E": 3,
      "G": 4,
    },
    "presence": {
      "A": 1,
      "B": 0.2438918433076174,
      "C": 0.6079095198861508,
      "D": 0.3573140588719945,
      "E": 0.4976630963972735,
      "G": 0.41138491498764135,
    },
  },
}
`;

exports[`chord detection 1`] = `
[
  {
    "chordNotes": Set {
      "G",
      "A",
      "E",
      "C",
      "D",
    },
    "chords": [
      "C6/9/G",
      "D11/G",
      "D9sus4/G",
      "Am7add11/G",
    ],
    "distribution": {
      "ticks": {
        "A": 369,
        "B": 15,
        "C": 123,
        "D": 49,
        "E": 175,
        "G": 144,
      },
      "time": {
        "A": 0.76875,
        "B": 0.03125,
        "C": 0.25625,
        "D": 0.10208333333333333,
        "E": 0.3645833333333333,
        "G": 0.3,
      },
    },
    "durationTicks": 480,
    "notes": Set {
      "G",
      "A",
      "E",
      "C",
      "D",
      "B",
    },
    "startTicks": 0,
    "uniqueNotes": [],
    "unit": "beat",
    "unitNumber": 0,
  },
  {
    "chordNotes": Set {
      "A",
      "E",
      "C",
      "B",
      "D",
    },
    "chords": [],
    "distribution": {
      "ticks": {
        "A": 302,
        "B": 82,
        "C": 232,
        "D": 55,
        "E": 130,
        "G": 41,
      },
      "time": {
        "A": 0.6291666666666667,
        "B": 0.17083333333333334,
        "C": 0.48333333333333334,
        "D": 0.11458333333333333,
        "E": 0.2708333333333333,
        "G": 0.08541666666666667,
      },
    },
    "durationTicks": 480,
    "notes": Set {
      "A",
      "E",
      "C",
      "G",
      "B",
      "D",
    },
    "startTicks": 480,
    "uniqueNotes": [],
    "unit": "beat",
    "unitNumber": 1,
  },
  {
    "chordNotes": Set {
      "G",
      "E",
      "A",
      "C",
    },
    "chords": [
      "C6/G",
      "Am7/G",
    ],
    "distribution": {
      "ticks": {
        "A": 159,
        "C": 85,
        "D": 19,
        "E": 217,
        "G": 137,
      },
      "time": {
        "A": 0.33125,
        "C": 0.17708333333333334,
        "D": 0.03958333333333333,
        "E": 0.45208333333333334,
        "G": 0.28541666666666665,
      },
    },
    "durationTicks": 480,
    "notes": Set {
      "G",
      "E",
      "A",
      "C",
      "D",
    },
    "startTicks": 960,
    "uniqueNotes": [],
    "unit": "beat",
    "unitNumber": 2,
  },
  {
    "chordNotes": Set {
      "A",
      "C",
      "D",
      "G",
      "B",
    },
    "chords": [],
    "distribution": {
      "ticks": {
        "A": 197,
        "B": 167,
        "C": 310,
        "D": 200,
        "E": 47,
        "G": 104,
      },
      "time": {
        "A": 0.41041666666666665,
        "B": 0.34791666666666665,
        "C": 0.6458333333333334,
        "D": 0.4166666666666667,
        "E": 0.09791666666666667,
        "G": 0.21666666666666667,
      },
    },
    "durationTicks": 480,
    "notes": Set {
      "E",
      "A",
      "C",
      "D",
      "G",
      "B",
    },
    "startTicks": 1440,
    "uniqueNotes": [],
    "unit": "beat",
    "unitNumber": 3,
  },
  {
    "chordNotes": Set {},
    "chords": [],
    "distribution": {
      "ticks": {
        "E": 3,
      },
      "time": {
        "E": 0.00625,
      },
    },
    "durationTicks": 480,
    "notes": Set {
      "E",
    },
    "startTicks": 1920,
    "uniqueNotes": [],
    "unit": "beat",
    "unitNumber": 4,
  },
]
`;

exports[`parseMidi() should parse to a known format 1`] = `
{
  "timeSignature": {
    "denominator": 4,
    "numerator": 4,
  },
  "timings": {
    "durationTicks": 479,
    "microsecondsPerBeat": 517241,
    "ticksPerBeat": 480,
  },
  "tracks": [
    {
      "durationTicks": 479,
      "name": "Pianoteq",
      "notes": [
        {
          "durationTicks": 35,
          "noteName": [
            "A",
            "A",
          ],
          "noteNameWithOctave": [
            "A3",
            "A3",
          ],
          "noteNumber": 57,
          "startTicks": 39,
          "uuid": "removed",
        },
        {
          "durationTicks": 55,
          "noteName": [
            "D",
            "D",
          ],
          "noteNameWithOctave": [
            "D4",
            "D4",
          ],
          "noteNumber": 62,
          "startTicks": 177,
          "uuid": "removed",
        },
        {
          "durationTicks": 43,
          "noteName": [
            "Bb",
            "A#",
          ],
          "noteNameWithOctave": [
            "Bb3",
            "A#3",
          ],
          "noteNumber": 58,
          "startTicks": 273,
          "uuid": "removed",
        },
        {
          "durationTicks": 34,
          "noteName": [
            "A",
            "A",
          ],
          "noteNameWithOctave": [
            "A2",
            "A2",
          ],
          "noteNumber": 45,
          "startTicks": 393,
          "uuid": "removed",
        },
        {
          "durationTicks": 18,
          "noteName": [
            "G",
            "G",
          ],
          "noteNameWithOctave": [
            "G4",
            "G4",
          ],
          "noteNumber": 67,
          "startTicks": 479,
          "uuid": "removed",
        },
      ],
    },
  ],
}
`;
